xpdan/callbacks.py

```diff
@@ -7,3 +17,145 @@ def start(self, doc):
 
     def stop(self, doc):
         print('FINISH ANALYSIS ON {}'.format(doc.get('run_start', 'NA')))
+
+
+class ReturnCallback(CallbackBase):
+    def __call__(self, name, doc):
+        ret = getattr(self, name)(doc)
+        if ret:
+            return name, ret
+        return name, doc
+
+
+class Retrieve(ReturnCallback):
+    def __init__(self, handler_reg, root_map=None):
+        if root_map is None:
+            root_map = {}
+        if handler_reg is None:
+            handler_reg = {}
+        self.resources = None
+        self.handlers = None
+        self.datums = None
+        self.root_map = root_map
+        self.handler_reg = handler_reg
+
+    def start(self, doc):
+        self.resources = {}
+        self.handlers = {}
+        self.datums = {}
+
+    def resource(self, resource):
+        self.resources[resource['uid']] = resource
+        handler = self.handler_reg[resource['spec']]
+
+        key = (str(resource['uid']), handler.__name__)
+
+        kwargs = resource['resource_kwargs']
+        rpath = resource['resource_path']
+        root = resource.get('root', '')
+        root = self.root_map.get(root, root)
+        if root:
+            rpath = os.path.join(root, rpath)
+        ret = handler(rpath, **kwargs)
+        self.handlers[key] = ret
+
+    def datum(self, doc):
+        self.datums[doc['datum_id']] = doc
+
+    def retrieve_datum(self, datum_id):
+        doc = self.datums[datum_id]
+        resource = self.resources[doc['resource']]
+        handler_class = self.handler_reg[resource['spec']]
+        key = (str(resource['uid']), handler_class.__name__)
+        return self.handlers[key](**doc['datum_kwargs'])
+
+    def fill_event(self, event, fields=True, inplace=True):
+        if fields is True:
+            fields = set(event['data'])
+        elif fields is False:
+            # if no fields, we got nothing to do!
+            # just return back as-is
+            return event
+        elif fields:
+            fields = set(fields)
+
+        if not inplace:
+            ev = _sanitize(event)
+            ev = copy.deepcopy(ev)
+        else:
+            ev = event
+        data = ev['data']
+        filled = ev['filled']
+        for k in set(data) & fields:
+            # Try to fill the data
+            try:
+                print('hiw')
+                print(k, data[k])
+                v = self.retrieve_datum(data[k])
+                print(k, data[k], v)
+                data[k] = v
+                filled[k] = True
+            # If retrieve fails keep going
+            except (ValueError, KeyError) as e:
+                raise e
+        return event
+
+    def event(self, doc):
+        ev = self.fill_event(doc, inplace=False)
+        return ev
+
+
+class ExportCallback(Retrieve):
+    """Callback to copy data to a new root"""
+
+    def __init__(self, new_root, handler_reg, root_map=None):
+        # TODO: fix this, since it is a dirty hack which only works for ADTIFF
```

Comment:
We need something, maybe a meta handler which looks like a handler but always returns the file path for that datum. For most of the specs this would call `file_list` but for tiffs we need to use the `PathOnlyHandler`